packages/accounts/src/light-account/account.ts

```diff
@@ -192,148 +170,74 @@ export async function createLightAccount<
-          },
-          getDummySignature: () => {
-            return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
-          },
-          encodeUpgradeToAndCall,
-        })
-      : _entryPoint.version === "0.7.0"
-      ? await toSmartContractAccount({
-          transport,
-          chain,
-          entryPoint: _entryPoint as EntryPointDef<"0.7.0">,
-          accountAddress: address,
-          source: "LightAccount",
-          getAccountInitCode,
-          encodeExecute: async ({ target, data, value }) => {
-            return encodeFunctionData({
-              abi: LightAccountAbi,
-              functionName: "execute",
-              args: [target, value ?? 0n, data],
-            });
-          },
-          encodeBatchExecute: async (txs) => {
-            const [targets, values, datas] = txs.reduce(
-              (accum, curr) => {
-                accum[0].push(curr.target);
-                accum[1].push(curr.value ?? 0n);
-                accum[2].push(curr.data);
-
-                return accum;
-              },
-              [[], [], []] as [Address[], bigint[], Hex[]]
-            );
-            return encodeFunctionData({
-              abi: LightAccountAbi,
-              functionName: "executeBatch",
-              args: [targets, values, datas],
-            });
-          },
-          signUserOperationHash: async (uoHash: Hex) => {
-            return signer.signMessage({ raw: uoHash });
-          },
-          async signMessage({ message }) {
-            const version = await getLightAccountVersion(account!);
-            switch (version) {
-              case "v1.0.1":
-                return signer.signMessage(message);
-              case "v1.0.2":
-                throw new Error(
-                  `Version ${version} of LightAccount doesn't support 1271`
-                );
-              default:
-                return signWith1271Wrapper(hashMessage(message));
-            }
-          },
-          async signTypedData(params) {
-            const version = await getLightAccountVersion(account!);
-            switch (version) {
-              case "v1.0.1": {
-                return signer.signTypedData(
-                  params as unknown as SignTypedDataParameters
-                );
-              }
-              case "v1.0.2":
-                throw new Error(
-                  `Version ${version} of LightAccount doesn't support 1271`
-                );
-              default:
-                return signWith1271Wrapper(hashTypedData(params));
-            }
-          },
-          getDummySignature: () => {
-            return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
-          },
-          encodeUpgradeToAndCall,
-        })
-      : undefined;
+  const account = await toSmartContractAccount({
```

Comment:
this will not work. We need to specifically return the strong typed account by checking the version and explicitly casting to the typed account. This is the trick that is used to the devs just "optionally" pass in entry point def, and they never have to strongly type `createXXAccount` methods. The returned type of the account will be strongly typed to the entry point version just based on whether the entrypoint def is 0.6 or 0.7.

You can refer to how I cleaned up/refactored this code to make it nicer here https://github.com/alchemyplatform/aa-sdk/pull/543 (optimize branch pr)

cc: @dphilipson @dancoombs

Reply:
this works just fine: https://github.com/alchemyplatform/aa-sdk/pull/544 and allows the overloaded method to support any of the entrypoints that the account supports.

I had to update the generics to not use the `EntryPointParameter` and just use `EntryPointDef` plus add the optional `TEntryPointVersion` to light account in a few more places

Reply:
the reason #544 works is because of the defaults on the generic being the union + the overloading which results in the implementation operating on the catch all. with that, you don't really need to do the type casting to a specific LightAccount type because typescript will correctly infer the output based on the `entryPoint` parameter passed into `createLightAccount`